### ISSUE TITLE
Inheriting from instantiated generic class causes mocking to fail

### ISSUE THREAD

**Issue by jhance** [python/typing#250]
My reproduction (python 2)

```
import mock

from typing import Generic, TypeVar

T = TypeVar('T')

class Bar(Generic[T]):
    def __init__(self):
        pass

    def do_bar(self):
        print("bar")

class Foo(Bar[int]):
    def do_foo(self):
        self.do_bar()

@mock.patch.object(Bar, 'do_bar')
def mocked(mock_do_bar):
    mock_do_bar.return_value = None
    foo = Foo()
    foo.do_foo()

mocked()
```

will unexpectedly print `bar`.

The inheritance here is crucial.

**Comment by gvanrossum** [comment]
(PS try to trim trailing whitespace when copy/pasting text into github. It causes spurious horizontal scroll bars to appear and weird line breaks. I cleaned it up for you this time.)

**Comment by jhance** [comment]
Oh, I did not realize I somehow accrued trailing whitespace. Must have been because I pastebin'd it to Jukka then copied from there.

**Comment by gvanrossum** [comment]
Anyway, I do think this is a bug, but I don't know how to fix it. Looking at the MRO for Foo, we have:
```python
[__main__.Foo,
 __main__.Bar<~T>[int],
 __main__.Bar<~T>,
 typing.Generic<~T>,
 typing.Generic,
 <type 'object'>]
```
I think the double occurrence of Bar may be the problem -- you're patching the original, but the MRO search finds the unpatched copy in Bar[int].

I think here's a workaround:
```python
...
BarInt = Bar[int]

class Foo(BarInt):
    def do_foo(self):
        self.do_bar()

@mock.patch.object(BarInt, 'do_bar')
def mocked(mock_do_bar):
...
```

**Comment by jhance** [comment]
I see.

Well, a workaround is good enough for now. Thanks!

**Comment by gvanrossum** [comment]
Sorry, this wasn't fixed by #289.

**Comment by jhance** [comment]
Thanks :D
